topology/probes/opencontrail/routing_table.go

```diff
@@ -0,0 +1,339 @@
+			mapper.addRoute(a.route.VrfId, ocRoute)
+			vrfId = a.route.VrfId
+		} else if a.action == DelRoute {
+			ocRoute := OpenContrailRoute{
+				Protocol: OpenContrailRouteProtocol,
+				Prefix:   fmt.Sprintf("%s/%d", a.route.Address, a.route.Prefix),
+				Family:   a.route.Family,
+				NhId:     a.route.NhId}
+			mapper.delRoute(a.route.VrfId, ocRoute)
+			vrfId = a.route.VrfId
+		} else if a.action == AddInterface {
+			mapper.addInterface(a.intf.VrfId, a.intf.InterfaceUUID)
+			vrfId = a.intf.VrfId
+
+		} else if a.action == DelInterface {
+			var err error
+			if vrfId, err = mapper.deleteInterface(a.intf.InterfaceUUID); err != nil {
+				continue
+			}
+		}
+		mapper.onRouteChanged(vrfId)
+	}
+}
+
+func (mapper *OpenContrailProbe) getOrCreateRoutingTable(vrfId int) *RoutingTable {
+	vrf, exists := mapper.routingTables[vrfId]
+	if !exists {
+		logging.GetLogger().Debugf("Creating a new VRF with ID %d", vrfId)
+		itfs := []string{}
+		vrf = &RoutingTable{InterfacesUUID: itfs}
+		mapper.routingTables[vrfId] = vrf
+		err := mapper.vrfInit(vrfId)
+		if err != nil {
+			logging.GetLogger().Errorf(err.Error())
+		}
+	}
+	return vrf
+}
+
+func (mapper *OpenContrailProbe) addInterface(vrfId int, interfaceUUID string) {
+	vrf := mapper.getOrCreateRoutingTable(vrfId)
+	logging.GetLogger().Debugf("Appending interface %s to VRF %d...", interfaceUUID, vrfId)
+	vrf.InterfacesUUID = append(vrf.InterfacesUUID, interfaceUUID)
+}
+
+func (mapper *OpenContrailProbe) OnInterfaceAdded(vrfId int, interfaceUUID string) {
+	mapper.routingTableUpdaterChan <- RoutingTableUpdate{action: AddInterface, intf: interfaceUpdate{InterfaceUUID: interfaceUUID, VrfId: vrfId}}
+}
+
+// deleteInterface removes interfaces from Vrf. If a Vrf no longer has
+// any interfaces, this Vrf is removed.
+func (mapper *OpenContrailProbe) deleteInterface(interfaceUUID string) (vrfId int, err error) {
+	var found bool
+	for k, vrf := range mapper.routingTables {
+		for idx, intf := range vrf.InterfacesUUID {
+			if intf == interfaceUUID {
+				logging.GetLogger().Debugf("Delete interface %s from VRF %d", interfaceUUID, k)
+				vrf.InterfacesUUID[idx] = vrf.InterfacesUUID[len(vrf.InterfacesUUID)-1]
+				vrf.InterfacesUUID = vrf.InterfacesUUID[:len(vrf.InterfacesUUID)-1]
+				found = true
+				break
+			}
+		}
+		if found {
+			if len(vrf.InterfacesUUID) == 0 {
+				logging.GetLogger().Debugf("Delete VRF %d", k)
+				delete(mapper.routingTables, k)
+			}
+		}
+	}
+	return 0, errors.New("No VrfId was found")
+}
+
+func (mapper *OpenContrailProbe) OnInterfaceDeleted(interfaceUUID string) {
+	mapper.routingTableUpdaterChan <- RoutingTableUpdate{action: DelInterface, intf: interfaceUpdate{InterfaceUUID: interfaceUUID}}
+}
+
+// onRouteChanged writes the Contrail routing table into the
+// Contrail.RoutingTable metadata attribute.
+func (mapper *OpenContrailProbe) onRouteChanged(vrfId int) {
+	mapper.graph.Lock()
+	defer mapper.graph.Unlock()
+
+	vrf := mapper.getOrCreateRoutingTable(vrfId)
```

Comment:
this could/should be done before locking no ?

Reply:
yes